### PR TITLE
Added support for custom RateLimitKeyer @Bean

### PR DESCRIPTION
**Pull request by roxspring**
Allows the RateLimitKey to be customised to application requirements beyond those offered through configuration.

### CODE REVIEW

**Comment by marcosbarbero** [comment]
Wow that's a nice one!
I'll check it in details tomorrow morning and probably release a new version right after!

File: spring-cloud-zuul-ratelimit-core/src/main/java/com/marcosbarbero/cloud/autoconfigure/zuul/ratelimit/config/RateLimitKeyer.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package com.marcosbarbero.cloud.autoconfigure.zuul.ratelimit.config;
+
+import com.marcosbarbero.cloud.autoconfigure.zuul.ratelimit.config.properties.RateLimitProperties;
+import org.springframework.cloud.netflix.zuul.filters.Route;
+
+import javax.servlet.http.HttpServletRequest;
+
+public interface RateLimitKeyer {
```
Comment (marcosbarbero) [inline spring-cloud-zuul-ratelimit-core/src/main/java/com/marcosbarbero/cloud/autoconfigure/zuul/ratelimit/config/RateLimitKeyer.java]: I think a name like `KeyGenerator` or `RateLimitKeyGenerator` will fit better.
WDYT?

Comment (roxspring) [reply spring-cloud-zuul-ratelimit-core/src/main/java/com/marcosbarbero/cloud/autoconfigure/zuul/ratelimit/config/RateLimitKeyer.java]: I went through several combinations of method/class/package names to use and wasn't particularly happy with any of them.

`RateLimit` seems to be a prefix used so will update to `RateLimitKeyGenerator`.

**Comment by marcosbarbero** [comment]
@roxspring here it is [v1.3.2.RELEASE](https://github.com/marcosbarbero/spring-cloud-zuul-ratelimit/releases/tag/v1.3.2.RELEASE).

Thanks a lot for your contribution.

I was thinking, giving that new approach maybe the `Policy.types` shouldn't be an `enum` anymore and now it should be a List of `String` otherwise if it's configured with something diff than `ORIGIN, USER, URL` it can't be used in a custom `RateLimitKeyGenerator`.

WDYT?

**Comment by roxspring** [comment]
Thanks for the quick release!

Using String as the policy type does sound useful if people need to be able to do basic configuration of their custom key generator. That said, if people are coding up a customisation then I don’t know how much they need configuration too. 

For my use case I want to use a default policy and take control so that I can (a) have a server wide key ignoring the route, and (b) mix in information from an authorization bearer token without needing full authn for the zuul proxy. So for my use case I’ll likely be ignoring most of the configuration anyway!

**Comment by marcosbarbero** [comment]
@roxspring this bearer token use case is quite interesting, but that part of the code doesn't cover it for you?

```
if (types.contains(Type.USER)) {
                joiner.add(request.getUserPrincipal() != null ? request.getUserPrincipal().getName() : ANONYMOUS_USER);
}
```

**Comment by marcosbarbero** [comment]
I got it, thanks for sharing!

**Pull request merged**